internal/core/item_tree.rs

```diff
@@ -312,21 +313,30 @@ impl ItemRc {
         r.upgrade()?.parent_item()
     }
 
-    // FIXME: This should be nicer/done elsewhere?
+    /// Returns true if this item is visible from the root of the item tree. Note that this will return
+    /// false for `Clip` elements, even if they are "visible".
```

Comment:
Will it? All Clip elements? Why is that?

Reply:
Maybe I'm missing something (I tried to cover that also in the test I added in search_api.rs), but the old code did this:

```rust
    pub fn is_visible(&self) -> bool {
        let item = self.borrow();
        let is_clipping = crate::item_rendering::is_clipping_item(item);
        let geometry = self.geometry();
        if is_clipping && (geometry.width() <= 0.01 as _ || geometry.height() <= 0.01 as _) {
            return false;
        }
        if let Some(parent) = self.parent_item() {
            parent.is_visible()
        } else {
            true
        }
    }
```

That suggests to me that `is_clipping_item()` returns true for `Clip` elements that are enabled for clipping (clip true) and the visible pass gives them the null width/height. So I think that'll return false.

The new code should return an empty clip rectangle from the call to `absolute_clip_rect_and_geometry` because of this:

```rust
      if crate::item_rendering::is_clipping_item(item) {
            clip = geometry.intersection(&clip).unwrap_or_default();
        }
```

And then the intersection of itself should be empty.

Should I add a direct unit test in here perhaps?

Reply:
A test for that would be nice.

So if you have a 
```slint
outer := Rectangle { 
   clip: true;  
   inner := Rectangle { x: parent.width; background: blue; } 
}
```
Then the `Clip` item for that outer rectangle would still be considered visible. 
But the inner Rectangle would not;

I think either behavior is fine. It's just that the documention is a bit strange.
I think what you mean is that if an item has a size of 0x0 it will not be considered visible (because the intersection is empty)

Reply:
> Then the Clip item for that outer rectangle would still be considered visible.
> But the inner Rectangle would not;

The `Clip` item would not be considered visible (the function returns `false`). AFAICS the old code did that as per this condition:

```rust
        if is_clipping && (geometry.width() <= 0.01 as _ || geometry.height() <= 0.01 as _) {  
            return false;  
        }  
```

and the new code should do the same by intersection.